### PR TITLE
-P acceptance of email address

### PR DESCRIPTION
**Pull request by AndreaLanfranchi**
Allow usage of email address in nanopool
Addresses
#1004
#937 
#904

_Email MUST be Path part of URI_ 
should be 
Email **CAN** be placed unencoded in path part

eg. -P
stratum+tcp://ethaddress[.worker][:password]@eth-eu1.nanopool.org:9999[/emailaddress]

### CODE REVIEW

**Comment by AndreaLanfranchi** [comment]
New valid syntaxes for nanopool 

-P stratum1+tcp://wallet.worker@eth-eu1.nanopool.org:9999/john.doe@gmail.com
-P stratum1+tcp://wallet@eth-eu1.nanopool.org:9999/workername/john.doe@gmail.com

Generally speaking everything in uri.Path() is appended to uri.User()

**Comment by chfast** [comment]
I don't think this is needed. You can include email address in worker name: `john.doe%40gmail.com`.

**Comment by AndreaLanfranchi** [comment]
For the sake of semplicity I think it mandatory.
Avg users do not have idea about how to encode email address plus, 
leveraging path part we can endorse official nanopool rules with slashes


Sent from mobile. Apologies for brevity and typos.

In data 2 maggio 2018 21:48:19 Paweł Bylica <notifications@github.com> ha 
scritto:
I don't think this is needed. You can include email address in worker name: 
john.doe%40gmail.com.

**Comment by AndreaLanfranchi** [comment]
Also consider there are additional adjustments

Sent from mobile. Apologies for brevity and typos.

In data 2 maggio 2018 21:48:19 Paweł Bylica <notifications@github.com> ha 
scritto:
I don't think this is needed. You can include email address in worker name: 
john.doe%40gmail.com.

**Comment by chfast** [comment]
Maybe, new changes should not go to release/0.14 branch unless they are strict fixes.

Moreover, it looks like including worker name in the pool URL is a dead end. I think we should separate these 2 things especially when you want to provide multiple pool URLs and single worker name.

**Comment by AndreaLanfranchi** [comment]
As URI is a feature of 0.14 thought this was a fix for people complaining 
they could not use ethminer on nanopool after implementation of -P


Sent from mobile. Apologies for brevity and typos.

In data 2 maggio 2018 22:56:56 Paweł Bylica <notifications@github.com> ha 
scritto:
Maybe, new changes should not go to release/0.14 branch unless they are 
strict fixes.
Moreover, it looks like including worker name in the pool URL is a dead 
end. I think we should separate these 2 things especially when you want to 
provide multiple pool URLs and single worker name.

**Comment by chfast** [comment]
> As URI is a feature of 0.14 thought this was a fix for people complaining
> they could not use ethminer on nanopool after implementation of -P

Yes, I know. But there should be also the old way of passing arguments.

I think we should rethink what syntax we want. This looks `stratum+tcp://ethaddress[.worker][:password]@eth-eu1.nanopool.org:9999[/emailaddress]` quite complex. The worker name consists of 4 pieces in different parts of the URI. That's not consistent.

I believe something like this would be easier to users:

```
--worker-name ethaddress.worker/email@example.com:password --pool stratum+tcp://eth-eu1.nanopool.org:9999 --pool stratum+tls://eth-eu2.nanopool.org:9999 
```

Pros:
1. You don't have to repeat your worker name for each pool URL.
2. "Public" options are separated for "private" options.

I'm not sure `--worker-name` is the best name, can be something else like `--user`. We can ask users about it.

**Comment by SnowLeopard71** [comment]
The _complex_ way would appear to be the technically the correct way, according to the URI specification, RFC 3986.
[https://en.wikipedia.org/wiki/Uniform_Resource_Identifier](url)
[https://datatracker.ietf.org/doc/rfc3986/](url)
Although I do not see why multiple syntaxes cannot be supported.

**Comment by AndreaLanfranchi** [comment]
Old way is completely preserved.
This is not a replacement but an adjustment in favour of usability.

For the rest I do not see the point of having a global worker name. I 
perceive it as a regression to old -O parameter.
-P has introduced great flexibility but has the drawback of URI structure 
where the authority part of the URI itself badly couple with the fantasy 
pools use for their logins.

I reiterate this is not a replacement but a fix for a feature which many 
perceive as non functional.

Anyway ...


Sent from mobile. Apologies for brevity and typos.

In data 2 maggio 2018 23:53:05 Paweł Bylica <notifications@github.com> ha 
scritto:
As URI is a feature of 0.14 thought this was a fix for people complaining
they could not use ethminer on nanopool after implementation of -PYes, I 
know. But there should be also the old way of passing arguments.
I think we should rethink what syntax we want. This looks 
stratum+tcp://ethaddress[.worker][:password]@eth-eu1.nanopool.org:9999[/emailaddress] 
quite complex. The worker name consists of 4 pieces in different parts of 
the URI. That's not consistent.
I believe something like this would be easier to users:
--worker-name ethaddress.worker/email@example.com:password --pool 
stratum+tcp://eth-eu1.nanopool.org:9999 --pool 
stratum+tls://eth-eu2.nanopool.org:9999
Pros:
You don't have to repeat your worker name for each pool URL.
"Public" options are separated for "private" options.
I'm not sure --worker-name is the best name, can be something else like 
--user. We can ask users about it.

**Comment by smurfy** [comment]
I think the main benefit of ``-P``  is that its self sufficient and does not rely on any other parameter for the pool connection. I would like to keep it that way.

**Comment by AndreaLanfranchi** [comment]
-P is defacto an URI parser.
I see pools where the separator for worker name is plus sign or slash 
instead of dot.

We may decide to rigidly honor RFC specs (in a world where there is no pool 
which correctly implement jsonrpc !!) and force users to encode the 
authority part of URI (there will be issues opened again and again) ... Or 
... Just let this workaround do a dirty job. It does not break anything 
and, indeed, not a new feature.

Nevertheless this tweak is very similar to the one for ethminer to work 
with eth-proxy. Substantially no stratum server will ever use path part of URI

But I resign the arms on this matter as I generally prefer usability 
instead of purity

You decide

Sent from mobile. Apologies for brevity and typos.

In data 3 maggio 2018 00:15:17 Philipp Andreas <notifications@github.com> 
ha scritto:
I think the main benefit of -P is that its self sufficient and does not 
rely on any other parameter for the pool connection. I would like to keep 
it that way.

**Comment by chfast** [comment]
How about we move all userdata to the path part of the URL?

```
stratum+tcp://eth-eu1.nanopool.org:9999/0x8FB24C5b5a75887b429d886DBb57fd053D4CF3a2.workername/email@example.com:password
```

**Comment by AndreaLanfranchi** [comment]
More work, rewrite the parser to identify password (if any), break 
authority part of URI and eventually compatibility broken.
I also reckon URI parser inherits from another project.
Would do that only in junction with implementation of boost program args


Sent from mobile. Apologies for brevity and typos.

In data 3 maggio 2018 00:55:06 Paweł Bylica <notifications@github.com> ha 
scritto:
How about we move all userdata to the path part of the URL?
stratum+tcp://eth-eu1.nanopool.org:9999/0x8FB24C5b5a75887b429d886DBb57fd053D4CF3a2.workername/email@example.com:password

**Comment by AndreaLanfranchi** [comment]
Also ...
Why not slashes like nanopool requires ?

stratum+tcp://eth-eu1.nanopool.org:9999/0x8FB24C5b5a75887b429d886DBb57fd053D4CF3a2.workername/email@example.com:password
stratum+tcp://eth-eu1.nanopool.org:9999/0x8FB24C5b5a75887b429d886DBb57fd053D4CF3a2/workername/email@example.com:password

Problem is pools are not compliant to any specification as, at present, 
there are none

Sent from mobile. Apologies for brevity and typos.

In data 3 maggio 2018 00:55:06 Paweł Bylica <notifications@github.com> ha 
scritto:
How about we move all userdata to the path part of the URL?
stratum+tcp://eth-eu1.nanopool.org:9999/0x8FB24C5b5a75887b429d886DBb57fd053D4CF3a2.workername/email@example.com:password

**Comment by ddobreff** [comment]
Getting more and more complex to assimilate.
Another way is to exclude just worker/rig from URI part --pool-worker(-PW) rig1/user@email.com --pool(-P) stratum+tcp:// or none, stratum+ssl or ssl://wallet@pool.address:pool.port 
Either way it got complex since it was introduced.

**Comment by jean-m-cyr** [comment]
I would favor keeping it all as it is, and add an optional --user parameter to specify the minerName/password content which would override whatever (if any) is specified in the -P parameters.

**Comment by AndreaLanfranchi** [comment]
>I believe something like this would be easier to users:
--worker-name ethaddress.worker/email@example.com:password --pool stratum+tcp://eth-eu1.nanopool.org:9999 --pool stratum+tls://eth-eu2.nanopool.org:9999 

It would be simpler if not breakes the degree of freedom -P introduced.
Actually each -P is autonomous: I can implement a backup (or more than one) pool where I mine with a completely different address or even on a completely different coin.
What about mining primarily on ETH and having as a backup UBIQ ? Your proposal would breach this freedom ?

**Comment by chfast** [comment]
Thanks for the summary @AndreaLanfranchi I really like it.

But there is still a small issue. The worker name is in authorization part, not in URL path. It will be difficult to move it there later on and keep backward compatibility. So maybe move it here in 0.14?

**Comment by AndreaLanfranchi** [comment]
>But there is still a small issue. The worker name is in authorization part, not in URL path. It will be difficult to move it there later on and keep backward compatibility. So maybe move it here in 0.14?

It's not ! Or at least not in a mandatory way.
URI parser inherits (I assume) from this https://github.com/cpp-netlib/uri
All it does is split the given URI in the corresponding parts as per this

```
scheme:[//[user[:password]@]host[:port]][/path][?query][#fragment]
           <---- authority -----------> 

```

So, actually we get "user" part valued with everything we need to accomplish pool's requirements to identify address, worker and eventually email.
As per the above scheme user *cannot* have slashes or `@` s "#" (see #1009)  (otherwise the parser assumes the host part to begin way before the proper part), you'd be obliged to _encode_ what you need in the user part.

Actually EthStratumClient does a job to try split user in two using dot as separator (hardcoded). What if any pool requires "+" ?

What PR does is, with no effort, add further separator "/" simply using the "path" thus sending a login which is nothing else than m_conn.User() + m_conn.Path() 

This way you can decide if you want to write

`-P stratum+tcp://0x01234567890123456789%2FWorkerName%2Femail%40example.com@ethermine.org:4444
`
or 

`
-P stratum+tcp://0x01234567890123456789@ethermine.org:4444/WorkerName/email@example.com
`

The first does not negate the second and viceversa.
Do you at least agree the second is way more readable ?

**Comment by chfast** [comment]
I only mean this:

The description of this PR states that
```
stratum+tcp://ethaddress[.worker][:password]@eth-eu1.nanopool.org:9999[/emailaddress]
```
so only email address is take from the path part, not 

```
stratum+tcp://0x01234567890123456789@ethermine.org:4444/WorkerName/email@example.com
```

**Comment by AndreaLanfranchi** [comment]
For the sake of clarity I could also add

`Example 4 : stratum1+tcp://0x01234567890123456789%2Fminer1%2Fjohn.doe%40gmail.com@nanopool.org:9999
`
but for sure it will issue "PANIC MODE ON" in noobs.

**Comment by AndreaLanfranchi** [comment]
Also all the issues referred in this PR where related to usage on Nanopool (which uses a / to identify email part).

File: libpoolprotocols/stratum/EthStratumClient.cpp [hunk 1]
```diff
@@ -368,13 +368,17 @@ void EthStratumClient::connect_handler(const boost::system::error_code& ec, tcp:
...
-		else
-			// Some marketing ?
-			m_worker = "ethminer " + std::string(ethminer_get_buildinfo()->project_version);
+		if (p != string::npos) {
+
+			user = m_conn.User().substr(0, p);
+			if (p + 1 <= m_conn.User().length())
+				m_worker = m_conn.User().substr(p + 1);
+			else
+				m_worker = "";
```
Comment (chfast) [inline libpoolprotocols/stratum/EthStratumClient.cpp]: Why do you clear the `m_worker` here?

**Comment by chfast** [comment]
Can I merge this?

**Comment by AndreaLanfranchi** [comment]
Yes !

**Comment by chfast** [comment]
Released.

**Pull request merged**